### ISSUE TITLE
Implicit conversions long long <--> int in linear algebra module

### ISSUE THREAD

**Issue by pelesh** [LLNL/hiop#150]
There are several implicit conversions from `int` to `long long` in the linear algebra module and related unit tests. This may cause some issues down the road. This has been confirmed to cause (at least part of the) problems in #103. 

A number of these conversions happens in sparse matrix methods, where arguments of the method are of `long long` type, while sparse matrix rows and columns are indexed by `int`s. None of these should be difficult to fix.

Perhaps it would be a good idea to add flag `-Wconversion` to the HiOp build configuration and ensure HiOp builds without  warnings. 

CC @cnpetra @nychiang @ashermancinelli

**Comment by ashermancinelli** [comment]
I believe we have some option for developer warnings. Are you suggesting enabling some warnings regardless of whether that option is enabled or not?

**Comment by cnpetra** [comment]
not sure if it;s clear why the interfaces are with `long long` (on short I would like to keep them like this for now)... the good news is these `long long` only store numbers that also fit in `int` and conversions from `long long` (not involving pointers) should not pose a problem

**Comment by pelesh** [comment]
> I believe we have some option for developer warnings. Are you suggesting enabling some warnings regardless of whether that option is enabled or not?

It looks to me we dropped `-Wconversion` flag from the developer's warnings. I think we only check for `float` <-> `double` conversiosn. I may be wrong.

**Comment by pelesh** [comment]
> not sure if it;s clear why the interfaces are with `long long` (on short I would like to keep them like this for now)... the good news is these `long long` only store numbers that also fit in `int` and conversions from `long long` (not involving pointers) should not pose a problem

For as long as we make sure all conversions are _explicit_, we should be fine. Potential issues may arise with mapping vectors:
```c++
  // keep track of the constraints indexes in the original, user's formulation
  long long *cons_eq_mapping_, *cons_ineq_mapping_; 
```

**Comment by ashermancinelli** [comment]
> > I believe we have some option for developer warnings. Are you suggesting enabling some warnings regardless of whether that option is enabled or not?
> 
> It looks to me we dropped `-Wconversion` flag from the developer's warnings. I think we only check for `float` <-> `double` conversiosn. I may be wrong.

No, you're right I didn't realize we dropped it. In the PR to address this I think we should add it back and turn `HIOP_DEVELOPER_MODE=ON` in the build script so we at least see these warnings in CI.

**Comment by cnpetra** [comment]
> > not sure if it;s clear why the interfaces are with `long long` (on short I would like to keep them like this for now)... the good news is these `long long` only store numbers that also fit in `int` and conversions from `long long` (not involving pointers) should not pose a problem
> 
> For as long as we make sure all conversions are _explicit_, we should be fine. Potential issues may arise with mapping vectors:
> 
> ```c++
>   // keep track of the constraints indexes in the original, user's formulation
>   long long *cons_eq_mapping_, *cons_ineq_mapping_; 
> ```
can you be more specific?

**Comment by nychiang** [comment]
We need to convert `long long` to `int` before passing it to RAJA functions in GPU.
Otherwise, unpredicted behavior may happen, as some compilers cannot convert `long long` to `int` on device, while others can.

For example, we need to do this conversion in function [row_max_abs_value](https://github.com/LLNL/hiop/blob/af9c15cc1610334c4e5003cd7ba2d653ceee66ca/src/LinAlg/hiopMatrixRajaDense.cpp#L1393-L1394) before the RAJA loop, as well as some other functions in [hiopMatrixRajaDense.cpp](https://github.com/LLNL/hiop/blob/dev-scaling/src/LinAlg/hiopMatrixRajaDense.cpp) . Otherwise, it will introduce run-time segment fault only marianas, while works fine on newell and ascent.

**Comment by pelesh** [comment]
I suggest a following "transitional" solution that would help us remove `-Wconversion` warnings in the near term:

Since `long long` is used only for (global) ordinals in HiOp, we could make a wholesale replacement from `long long` to something like `global_ordinal_type` and then set `using global_ordinal_type = int;` where appropriate. This is a quick fix to set all indices to `int` and remove conversion warnings. It will also allow one to get back to `long long`s easily, if needed.

**Comment by cnpetra** [comment]
fixed by PR#242
